### PR TITLE
PMM-6202 Automate tests for annotations adding

### PR DESCRIPTION
**Pull request by IvanaBrecherova**

### CODE REVIEW

File: tests/pages/pmmInventoryPage.js [hunk 1]
```diff
@@ -4,27 +4,30 @@ const assert = require('assert');
...
-    externalExporter: locate('td').withText('External exporter'),
+    pmmServicesSelector: '//li[contains(text(),"Services")]',
     postgresExporter: locate('td').withText('Postgres exporter'),
     postgresPgStatements: locate('td').withText('QAN PostgreSQL PgStatements'),
     postgresPgstatmonitor: locate('td').withText('QAN PostgreSQL Pgstatmonitor'),
+    proceedButton: '//span[contains(text(), "Proceed")]',
+    runningStatus: '//span[contains(text(), "RUNNING")]',
+    serviceIdLocatorPrefix: '//table//tr/td[4][contains(text(),"',
+    tableCheckbox: 'div[data-qa="select-row"]',
+    tableRow: '//tr[@data-qa="table-row"]',
```
Comment (puneet0191) [inline tests/pages/pmmInventoryPage.js:30]: since you are refactoring, why not use locate too?

File: tests/pages/pmmInventoryPage.js [hunk 2]
```diff
@@ -160,6 +163,10 @@ module.exports = {
     return serviceIdLocator;
   },
 
+  async getNodeName(nodeID) {
+    return await I.grabTextFrom(`//table//tr/td[2][contains(text(),"${nodeID}")]/following-sibling::td[2]`);
```
Comment (puneet0191) [inline tests/pages/pmmInventoryPage.js:167]: you could get this using API too

File: tests/verifyOSDashboards_test.js [hunk 3]
```diff
@@ -78,3 +85,43 @@ Data(nodes).Scenario(
     I.seeElement(dashboardPage.nodeSummaryDashboard.ptSummaryDetail.reportContainer);
   },
 );
+
+Scenario(
+  'PMM-T877 - Verify adding annotation for service name: pmm-server and node name: pmm-server-postgresql @nightly @dashboards',
```
Comment (puneet0191) [inline tests/verifyOSDashboards_test.js:90]: pmm-server is the node name, pmm-server-postgresql is the service name

File: tests/verifyOSDashboards_test.js [hunk 4]
```diff
@@ -78,3 +85,43 @@ Data(nodes).Scenario(
...
+    I.amOnPage(`${dashboardPage.postgresqlInstanceSummaryDashboard.url}`);
+    dashboardPage.waitForDashboardOpened();
+    await dashboardPage.applyFilter('Service Name', 'pmm-server-postgres');
+    dashboardPage.verifyAnnotationsLoaded(postgresAnnotation, 3);
+    I.seeElement(dashboardPage.annotationText(postgresAnnotation), 10);
+  },
+);
+
+Data(annotation).Scenario(
+  'PMM-T877 - Verify adding annotation for mongoDB @nightly @dashboards',
```
Comment (puneet0191) [inline tests/verifyOSDashboards_test.js:104]: since you have datatable, why use mongoDB in the scenario name?

File: tests/verifyOSDashboards_test.js [hunk 5]
```diff
@@ -78,3 +85,43 @@ Data(nodes).Scenario(
...
+    I, dashboardPage, pmmInventoryPage, grafanaAPI, inventoryAPI, current,
+  }) => {
+    const { annotationName } = current;
+
+    I.amOnPage(pmmInventoryPage.url);
+    const serviceName = await I.grabTextFrom(current.service);
+    const nodeID = await pmmInventoryPage.getNodeId(serviceName);
+    const nodeName = await inventoryAPI.getNodeName(nodeID);
+
+    await grafanaAPI.setAnnotation(annotationName, 'PMM-T877', nodeName.node_name, serviceName);
```
Comment (puneet0191) [inline tests/verifyOSDashboards_test.js:115]: why not return node_name directly?

File: tests/verifyOSDashboards_test.js [hunk 6]
```diff
@@ -1,7 +1,14 @@
+const { pmmInventoryPage, dashboardPage } = inject();
```
Comment (puneet0191) [inline tests/verifyOSDashboards_test.js:1]: why use OS Dashboards test for this?

**Review by puneet0191 — CHANGES_REQUESTED** [review]

File: tests/pages/pmmInventoryPage.js (hunk 2 above)
Comment (IvanaBrecherova) [reply tests/pages/pmmInventoryPage.js:167]: It's not used. I forgot to delete. Thanks! :)

**Comment by IvanaBrecherova** [comment]
https://pmm.cd.percona.com/blue/organizations/jenkins/pmm2-ui-test-nightly/detail/pmm2-ui-test-nightly/579/tests

File: tests/pages/api/annotationAPI.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+      text: annotationName,
+      tags: [tags],
+      node_name: nodeName,
+    };
+
+    const resp = await I.sendPostRequest('v1/management/Annotations/Add', body, headers);
+
+    assert.ok(
+      resp.status === 200,
+      `Failed to add annotation for service/s name: ${nodeName}. Response message is ${resp.data.message}`,
```
Comment (puneet0191) [inline tests/pages/api/annotationAPI.js:40]: please fix this, wrong assert

File: tests/pages/api/annotationAPI.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
+const { I } = inject();
+const assert = require('assert');
+
+/* eslint-disable consistent-return */
```
Comment (puneet0191) [inline tests/pages/api/annotationAPI.js:4]: why?

File: tests/pages/api/annotationAPI.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+      service_names: [serviceNames],
+    };
+
+    const resp = await I.sendPostRequest('v1/management/Annotations/Add', body, headers);
+
+    if (resp.status !== 200) {
+      return resp.status;
+    }
+
+    assert.ok(
```
Comment (puneet0191) [inline tests/pages/api/annotationAPI.js:22]: how would this assertion even run?

File: tests/pages/api/annotationAPI.js [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+      text: annotationName,
+      tags: [tags],
+      node_name: nodeName,
+      service_names: [serviceNames],
+    };
+
+    const resp = await I.sendPostRequest('v1/management/Annotations/Add', body, headers);
+
+    if (resp.status !== 200) {
+      return resp.status;
```
Comment (puneet0191) [inline tests/pages/api/annotationAPI.js:19]: why not keep it consistent, you could handle the response assert in tests too

File: tests/verifyAnnotations_test.js [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+    } else {
+      dashboardPage.verifyAnnotationsLoaded(annotationName, 1);
+    }
+
+    I.seeElement(dashboardPage.annotationText(annotationName), 10);
+  },
+);
+
+Data(annotation).Scenario(
+  'PMM-T878 - Verify not adding annotation with wrong node name @nightly @dashboards',
```
Comment (puneet0191) [inline tests/verifyAnnotations_test.js:58]: Please fix the text it is not really helping to understand the test purpose. 

Basically, you are checking if Annotations can not be added for wrong node names.

**Review by puneet0191 — CHANGES_REQUESTED** [review]

File: tests/pages/api/annotationAPI.js [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+    const body = {
+      text: annotationName,
+      tags: [tags],
+      node_name: nodeName,
+      service_names: [serviceNames],
+    };
+
+    const resp = await I.sendPostRequest('v1/management/Annotations/Add', body, headers);
+
+    if (resp.status !== 200) {
```
Comment (yurkovychv) [inline tests/pages/api/annotationAPI.js]: could you please explain why did you add this statement?

File: tests/pages/api/annotationAPI.js (hunk 8 above)
Comment (yurkovychv) [inline tests/pages/api/annotationAPI.js]: is it really needed here?

File: tests/pages/api/annotationAPI.js [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+      return resp.status;
+    }
+
+    assert.ok(
+      resp.status === 200,
+      `Failed to add annotation for service/s name: ${serviceNames}. Response message is ${resp.data.message}`,
+    );
+  },
+
+  async setAnnotationWithoutServiceName(annotationName, tags, nodeName) {
```
Comment (yurkovychv) [inline tests/pages/api/annotationAPI.js:19]: I think there is no need to duplicate methods. Please handle this in `setAnnotation` method

File: tests/pages/api/annotationAPI.js [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+
+    const resp = await I.sendPostRequest('v1/management/Annotations/Add', body, headers);
+
+    if (resp.status !== 200) {
+      return resp.status;
+    }
+
+    assert.ok(
+      resp.status === 200,
+      `Failed to add annotation for service/s name: ${serviceNames}. Response message is ${resp.data.message}`,
```
Comment (yurkovychv) [inline tests/pages/api/annotationAPI.js]:
```suggestion
      `Failed to add annotation for ${body}. Response message is ${resp.data.message}`,
```
This message will be more informative

File: tests/pages/pmmInventoryPage.js (hunk 1 above)
Comment (yurkovychv) [reply tests/pages/pmmInventoryPage.js:30]: @IvanaBrecherova please update other locators to use `locate` except `serviceIdLocatorPrefix`

File: tests/verifyAnnotations_test.js [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+    I.amOnPage(`${dashboardPage.postgresqlInstanceSummaryDashboard.url}`);
+    dashboardPage.waitForDashboardOpened();
+    await dashboardPage.applyFilter('Service Name', 'pmm-server-postgres');
+    dashboardPage.verifyAnnotationsLoaded(postgresAnnotation, 3);
+    I.seeElement(dashboardPage.annotationText(postgresAnnotation), 10);
+  },
+);
+
+Data(annotation).Scenario(
+  'PMM-T878 - Verify adding annotation specific dashboard @nightly @dashboards',
```
Comment (yurkovychv) [inline tests/verifyAnnotations_test.js:31]: Please move the first scenario to this Data table

File: tests/verifyAnnotations_test.js (hunk 11 above)
Comment (yurkovychv) [inline tests/verifyAnnotations_test.js]:
```suggestion
  'PMM-T878 - Verify user is not able to add an annotation for non-existing node @nightly @dashboards',
```

File: tests/verifyAnnotations_test.js [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+      dashboardPage.verifyAnnotationsLoaded(annotationName, 3);
+    } else {
+      dashboardPage.verifyAnnotationsLoaded(annotationName, 1);
+    }
+
+    I.seeElement(dashboardPage.annotationText(annotationName), 10);
+  },
+);
+
+Data(annotation).Scenario(
```
Comment (yurkovychv) [inline tests/verifyAnnotations_test.js:57]: I think there is no need to try different service names. Better to combine these 3 tests (wrong service name, wrong node name, without service name)

File: tests/verifyAnnotations_test.js [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+    const nodeName = await inventoryAPI.getNodeName(nodeID);
+
+    const errorNumber = await annotationAPI.setAnnotation('empty-service-name', 'PMM-T878', nodeName, '');
+
+    assert.ok(errorNumber === 400, `Annotation for ${nodeName} should not be added without service name`);
+  },
+);
+
+Scenario(
+  'PMM-T877 - Verify adding annotation for specific node @nightly @dashboards',
```
Comment (yurkovychv) [inline tests/verifyAnnotations_test.js:107]: This test can also be placed into add annotation data table

**Review by yurkovychv — CHANGES_REQUESTED** [review]

File: tests/pages/api/annotationAPI.js (hunk 8 above)
Comment (IvanaBrecherova) [reply tests/pages/api/annotationAPI.js]: It's unimportant after my fix. Deleted! :)

File: tests/pages/api/annotationAPI.js (hunk 9 above)
Comment (IvanaBrecherova) [reply tests/pages/api/annotationAPI.js]: bad implementation. Deleted! :)

File: tests/pages/api/annotationAPI.js [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+    const body = {
+      text: annotationName,
+      tags: [tags],
+      node_name: nodeName,
+      service_names: [serviceNames],
+    };
+
+    const resp = await I.sendPostRequest('v1/management/Annotations/Add', body, headers);
+
+    return resp.status;
```
Comment (yurkovychv) [inline tests/pages/api/annotationAPI.js:16]: We need an assert here

Comment (IvanaBrecherova) [reply tests/pages/api/annotationAPI.js:21]: @yurkovychv When I will add assert here, than I will not be able to return status for the tests when I am checking the status. Or I can add there again: `/* eslint-disable consistent-return */` and then I can have assert and return in if condition for the errors. What do you think? :)

Comment (yurkovychv) [reply tests/pages/api/annotationAPI.js:21]: what about an additional parameter in a function that will accept a status code to assert?

Comment (IvanaBrecherova) [reply tests/pages/api/annotationAPI.js:21]: Ok.

File: tests/pages/pmmInventoryPage.js [hunk 19]
```diff
@@ -4,35 +4,30 @@ const assert = require('assert');
...
+    pdphsqlServiceName: locate('td').withText('PDPGSQL'),
+    pmmAgentLocator: locate('td').withText('PMM Agent').after('table'),
+    pmmServerPostgresLocator: locate('td').withText('pmm-server-postgresql'),
+    pmmServicesSelector: locate('li').withText('Services'),
     postgresExporter: locate('td').withText('Postgres exporter'),
     postgresPgStatements: locate('td').withText('QAN PostgreSQL PgStatements'),
     postgresPgstatmonitor: locate('td').withText('QAN PostgreSQL Pgstatmonitor'),
-  },
-
-  verifyOldMySQLRemoteServiceIsDisplayed(serviceName) {
```
[the remark this answers is not in the document]
Comment (IvanaBrecherova) [reply tests/pages/pmmInventoryPage.js:30]: No.

File: tests/pages/api/annotationAPI.js [hunk 20]
```diff
@@ -0,0 +1,31 @@
+const { I } = inject();
+const assert = require('assert');
+
+module.exports = {
+  async setAnnotation(annotationName, tags, nodeName, serviceNames, statusCode) {
+    const headers = { Authorization: `Basic ${await I.getAuth()}` };
+    let body;
+
+    if (serviceNames !== null) {
```
[the remark this answers is not in the document]
Comment (IvanaBrecherova) [reply tests/pages/api/annotationAPI.js]: Doesn't work

**Comment by IvanaBrecherova** [comment]
Check it here: https://pmm.cd.percona.com/blue/organizations/jenkins/pmm2-ui-test-nightly/detail/pmm2-ui-test-nightly/592/pipeline

**Review by yurkovychv — APPROVED** [review]

**Review by puneet0191 — APPROVED** [review]
